dbms/src/Storages/StorageBuffer.cpp

```diff
@@ -178,7 +178,12 @@ BlockInputStreams StorageBuffer::read(
             LOG_WARNING(log, "Destination table " << backQuoteIfNeed(destination_database) << "." << backQuoteIfNeed(destination_table)
                 << " has no common columns with block in buffer. Block of data is skipped.");
         else
+        {
+            auto lock = destination->lockStructure(false, __PRETTY_FUNCTION__);
```

Comment:
Probably we should lock earlier (before the first call to `destination->hasColumn`)